concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/LifoBlockingLimiter.java

```diff
@@ -0,0 +1,214 @@
+            return this;
+        }
+
+        /**
+         * Function to derive the backlog timeout from the request context.  This allows timeouts
+         * to be set dynamically based on things like request deadlines. 
+         * @param supplier
+         * @param units
+         * @return
+         */
+        public Builder<ContextT> backlogTimeout(Function<ContextT, Long> mapper, TimeUnit units) {
+            this.maxBacklogTimeoutMillis = context -> units.toMillis(mapper.apply(context));
+            return this;
+        }
+
+        public LifoBlockingLimiter<ContextT> build() {
+            return new LifoBlockingLimiter<ContextT>(this);
+        }
+    }
+    
+    public static <ContextT> Builder<ContextT> newBuilder(Limiter<ContextT> delegate) {
+        return new Builder<ContextT>(delegate);
+    }
+    
+    private final Limiter<ContextT> delegate;
+    
+    private static class Event<ContextT> {
+        private volatile Optional<Listener> listener;
+        private final CountDownLatch latch = new CountDownLatch(1);
+        private ContextT context;
+        
+        public Event(ContextT context) {
+            this.context = context;
+        }
+
+        public boolean await(long timeout, TimeUnit unit) throws InterruptedException {
+            return latch.await(timeout, unit);
+        }
+        
+        public void set(Optional<Listener> listener) {
+            this.listener = listener;
+            latch.countDown();
+        }
+        
+    }
+    
+    /**
+     * Lock used to block and unblock callers as the limit is reached
+     */
+    private final Deque<Event<ContextT>> backlog = new LinkedList<>();
+    
+    private final AtomicInteger backlogCounter = new AtomicInteger();
+    
+    private final int backlogSize;
+    
+    private final Function<ContextT, Long> backlogTimeoutMillis;
+    
+    private final Object lock = new Object();
+    
+    private LifoBlockingLimiter(Builder<ContextT> builder) {
+        this.delegate = builder.delegate;
+        this.backlogSize = builder.maxBacklogSize;
+        backlogTimeoutMillis = builder.maxBacklogTimeoutMillis;
+    }
+    
+    private Optional<Listener> tryAcquire(ContextT context) {
+        // Try to acquire a token and return immediately if successful
+        final Optional<Listener> listener = delegate.acquire(context);
+        if (listener.isPresent()) {
+            return listener;
+        }
+
+        // Otherwise block until the limit has been released or we timeout 
+        if (backlogCounter.get() > this.backlogSize) {
+            return Optional.empty();
+        }
+        backlogCounter.incrementAndGet();
+        final Event<ContextT> event = new Event<>(context);
+        
+        try {
+            synchronized (lock) {
+                backlog.addFirst(event);
+            }
+            
+            if (!event.await(backlogTimeoutMillis.apply(context), TimeUnit.MILLISECONDS)) {
```

Comment:
Some comments here would be useful, maybe "If we wait on the latch but time out, we throttle the newest item in the backlog rather than the oldest." Also, shouldn't we be removing things from the end if we're timing out rather than the front?

Reply:
Oh wait, you're removing the first occurrence of the event not the first event, so in the worst case(and maybe even likely case) you'll be iterating O(N) elements in a linked list to find the item at the end and then remove it in O(1)? I might be worried about the CPU hit of this iteration in overload situations.

Reply:
You're also doing this iteration under a lock which scares me even more :)